cmd/remote-configuration/metrics/reporter.go

```diff
@@ -0,0 +1,209 @@
+	resource, resourceErr := otelresource.New(context.Background(),
+		otelresource.WithAttributes(
+			semconv.ServiceNameKey.String(agentType),
+			semconv.ServiceVersionKey.String(agentVersion),
+			semconv.ServiceInstanceIDKey.String(instanceId.String()),
+		),
+	)
+	if resourceErr != nil {
+		return nil, resourceErr
+	}
+
+	provider := sdkmetric.NewMeterProvider(
+		sdkmetric.WithResource(resource),
+		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(client)))
+
+	global.SetMeterProvider(provider)
+
+	reporter := &MetricReporter{
+		logger: logger,
+	}
+
+	reporter.done = make(chan struct{})
+
+	reporter.meter = global.Meter("opamp")
+
+	reporter.process, err = process.NewProcess(int32(os.Getpid()))
+	if err != nil {
+		return nil, fmt.Errorf("cannot query own process: %w", err)
+	}
+
+	// Create some metrics that will be reported according to OpenTelemetry semantic
+	// conventions for process metrics (conventions are TBD for now).
+	reporter.processCpuTime, err = reporter.meter.AsyncFloat64().Counter(
+		"process.cpu.time",
+	)
+	if err != nil {
+		return nil, fmt.Errorf("can't create process time metric: %w", err)
+	}
+	err = reporter.meter.RegisterCallback([]instrument.Asynchronous{reporter.processCpuTime}, reporter.processCpuTimeFunc)
+	if err != nil {
+		return nil, fmt.Errorf("can't create register callback: %w", err)
+	}
+	reporter.processMemoryPhysical, err = reporter.meter.AsyncInt64().Gauge(
+		"process.memory.physical_usage",
+	)
+	if err != nil {
+		return nil, fmt.Errorf("can't create memory metric: %w", err)
+	}
+	err = reporter.meter.RegisterCallback([]instrument.Asynchronous{reporter.processMemoryPhysical}, reporter.processMemoryPhysicalFunc)
+	if err != nil {
+		return nil, fmt.Errorf("can't register callback: %w", err)
+	}
+
+	reporter.counter, err = reporter.meter.SyncInt64().Counter("custom_metric_ticks")
+	if err != nil {
+		return nil, fmt.Errorf("can't register counter metric: %w", err)
+	}
+
+	reporter.meterShutdowner = func() { _ = provider.Shutdown(context.Background()) }
+
+	go reporter.sendMetrics()
+
+	return reporter, nil
+}
+
+func (reporter *MetricReporter) processCpuTimeFunc(c context.Context) {
+	times, err := reporter.process.Times()
+	if err != nil {
+		reporter.logger.Errorf("Cannot get process CPU times: %w", err)
+	}
+	reporter.processCpuTime.Observe(c, times.User, attribute.String("state", "user"))
+	reporter.processCpuTime.Observe(c, times.System, attribute.String("state", "system"))
+	reporter.processCpuTime.Observe(c, times.Iowait, attribute.String("state", "wait"))
+}
+
+func (reporter *MetricReporter) processMemoryPhysicalFunc(ctx context.Context) {
+	memory, err := reporter.process.MemoryInfo()
+	if err != nil {
+		reporter.logger.Errorf("Cannot get process memory information: %w", err)
+		return
+	}
+	reporter.processMemoryPhysical.Observe(ctx, int64(memory.RSS))
+}
+
+func (reporter *MetricReporter) sendMetrics() {
```

Comment:
Is this intended to have further function down the line?  I don't see that it's doing anything other than counting the number of 5s periods elapsed since it was started.  Actually collecting and sending metrics is governed by the `NewPeriodicReader()` created above.

Reply:
This code was from the example agent code from the opamp-go example client. I think we can remove it. This metrics code was the one thing I wasn't able to test well as part of this.

Reply:
tested this and everything worked after i moved to use the grpc client :D metrics were flowing through. I added a TODO to add more metrics in.